### DISCUSSION TITLE
OneDarkPro causes caret symbol in LuaLine when splitting window

### DISCUSSION THREAD

**Discussion by rnl16616** [olimorris/onedarkpro.nvim#111] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/31188607/195302191-8e131e71-31e6-468f-a390-87e419222b08.png)

Using NeoVim 0.8
When using a single window the status line shows correctly but when going into split window the status line is filled with the caret (^) symbol - see screenshot. This happens for both a vertical and horizontal split.

I first noticed it a week or so ago when I added in LSP. At first I thought it was just because I was struggling to setup LSP and DAP but now that is done I am still seeing the caret in split window mode. Even if I exclude LSP I still see the caret. I don't often use split windows so I am not sure whether the problem was going to NeoVim 0.8 or something else... If I use nvim -u NONE and split the window the status bar is fine. So it does seem that onedarkpro is the problem. Enabling onedarkpro causes Lualine to get these carets?

Is there any way to stop this behaviour from OneDarkPro? Or is there a way I can troubleshoot this? It seems there is nothing obviously in the configuration that would cause/allow this behaviour.

**onedarkpro.lua:**
local status_ok, odp = pcall(require, "onedarkpro")
if not status_ok then
    vim.notify("Failed to load onedarkpro", odp)
    return
end

odp.setup({
  dark_theme = "onedark", -- The default dark theme
  caching = false, -- Use caching for the theme?
  -- The path to the cache directory light_theme = "onelight", -- The default light theme
  cache_path = vim.fn.expand(vim.fn.stdpath("cache") .. "/onedarkpro/"),
  colors = {}, -- Override default by specifying colors for 'onelight' or 'onedark'
  plugins = { -- Override which plugin highlight groups are loaded
    -- See the Supported Plugins section for a list of available plugins
  },
  styles = { -- Choose from "bold,italic,underline"
      strings = "NONE", -- Style that is applied to strings.
      comments = "NONE", -- Style that is applied to comments
      keywords = "NONE", -- Style that is applied to keywords
      functions = "NONE", -- Style that is applied to functions
      variables = "NONE", -- Style that is applied to variables
      virtual_text = "NONE", -- Style that is applied to virtual text
  },
  options = {
      bold = false, -- Use the colorscheme's opinionated bold styles?
      italic = false, -- Use the colorscheme's opinionated italic styles?
      underline = false, -- Use the colorscheme's opinionated underline styles?
      undercurl = false, -- Use the colorscheme's opinionated undercurl styles?
      cursorline = false, -- Use cursorline highlighting?
      transparency = false, -- Use a transparent background?
      terminal_colors = false, -- Use the colorscheme's colors for Neovim's :terminal?
      window_unfocused_color = true, -- If window is out of focus, change background?
  }
})

**lualine.lua is very standard copy/paste with no customisation - just some options set:**
local status_ok, lualine = pcall(require,"lualine")
if not status_ok then
vim.notify("Failed to load lualine", lualine)
return
end

lualine.setup {
options = {
icons_enabled = true,
theme = 'onedark',
component_separators = { left = '', right = ''},
section_separators = { left = '', right = ''},
disabled_filetypes = {
statusline = {},
winbar = {},
},
ignore_focus = {},
always_divide_middle = true,
globalstatus = true,
refresh = {
statusline = 1000,
tabline = 1000,
winbar = 1000,
}
},
sections = {
lualine_a = {'mode'},
lualine_b = {'branch', 'diff', 'diagnostics'},
lualine_c = {'filename'},
lualine_x = {'buffers' },
lualine_y = {'encoding', 'filetype'},
lualine_z = {'location', 'progress'}
},
inactive_sections = {
lualine_a = {},
lualine_b = {},
lualine_c = {},
lualine_x = {},
lualine_y = {},
lualine_z = {}
},
tabline = {},
winbar = {},
inactive_winbar = {},
extensions = {}

**Comment by olimorris** [comment, marked as answer]
Think it may be related to this:

https://github.com/olimorris/onedarkpro.nvim/issues/109
https://github.com/olimorris/onedarkpro.nvim/issues/106

**Reply by rnl16616** [reply]
Thanks. As suggested, adding the following snippet to the setup() solved the problem:  
highlights = {StatusLine = { bg = "NONE", fg = "NONE" }},

Whilst I was playing with the settings - I also tried to reset the theme to another one rather than the default dark theme. It did not change anything? For example, using the theme = "onelight", I had expected it to change but it did not, similarly with using italic for comments.

I am using NeoVim 0.8 on Windows 11 via Windows Terminal 1.15x and FiraCode 6.2 fonts - is there a known problem here with switching the theme in Windows? (I appreciate that most people seem to be using linux and I am in a minority)

**Reply by olimorris** [reply]
Are you using caching?

**Reply by rnl16616** [reply]
No. Set to false currently. The setup() is basically using copy/paste:

odp.setup({
  dark_theme = "onedark", -- The default dark theme
  light_theme = "onelight", -- The default light theme
  theme = "onelight",
  caching = false, -- Use caching for the theme?
  -- The path to the cache directory light_theme = "onelight", -- The default light theme
  cache_path = vim.fn.expand(vim.fn.stdpath("cache") .. "/onedarkpro/"),
  colors = {}, -- Override default by specifying colors for 'onelight' or 'onedark'
  plugins = { -- Override which plugin highlight groups are loaded
    -- See the Supported Plugins section for a list of available plugins
  },
  highlights = {
      StatusLine = { bg = "NONE", fg = "NONE" }
  },
  styles = { -- Choose from "bold,italic,underline"
      strings = "NONE", -- Style that is applied to strings.
      comments = "italic", -- Style that is applied to comments
      keywords = "NONE", -- Style that is applied to keywords
      functions = "NONE", -- Style that is applied to functions
      variables = "NONE", -- Style that is applied to variables
      virtual_text = "NONE", -- Style that is applied to virtual text
  },
  options = {
      bold = false, -- Use the colorscheme's opinionated bold styles?
      italic = false, -- Use the colorscheme's opinionated italic styles?
      underline = false, -- Use the colorscheme's opinionated underline styles?
      undercurl = false, -- Use the colorscheme's opinionated undercurl styles?
      cursorline = false, -- Use cursorline highlighting?
      transparency = false, -- Use a transparent background?
      terminal_colors = false, -- Use the colorscheme's colors for Neovim's :terminal?
      window_unfocused_color = true, -- If window is out of focus, change background?
  }
})

**Reply by olimorris** [reply]
If you have everything set to false by default, I'd advise not to copy into your setup; it's just duplication of what's already in the `config.lua` file. Also, if your background is set to "light", which I'm guessing it is if you wish to use `onelight`, then you shouldn't need to use the `theme` option at all. The colorscheme will autoload `onelight` on your behalf.

Regardless of all of that, this setup should work. I've just tried it on my own config. Can you trim your setup file down to:

```lua
odp.setup({
  theme = "onelight"
})
vim.cmd("colorscheme onedarkpro")
```

and see if it works?

**Reply by rnl16616** [reply]
No it doesn't change the theme. I've modified the file to only include the changes. The background is normally dark I only set it to light as the change would be obvious. Evetually, I would actually like it set to the darkest option or onedark dark.

I actually continued to play around with your suggested setup:
odp.setup({
  theme = "onelight",
})
and whilst this did not switch the theme I noticed that splitting windows works again without the caret appearing. So not sure what is going on there but it looks like setting one particular option will trigger the caret appearing...

**Reply by olimorris** [reply]
Can you fill out an issue form and follow all the steps?

**Reply by rnl16616** [reply]
I used the minimal.lua with theme="onelight" and then "onedark_dark" and they both worked correctly. There must be something in my configuration that is the problem. I stopped reporting the issue as it is not clear that the problem is with onedarkpro.

I'll start from the minimal configuration and add things back in again until something breaks. Perhaps it is related to treesitter as I periodically (apparently randomly) get one error that is not resolved yet and is on my TODO list:
![image](https://user-images.githubusercontent.com/31188607/195352388-f49fc1cd-2cec-4637-a950-267bfcf8768e.png)

**Reply by olimorris** [reply]
I _always_ get problems with the damn `vim.so` parser. I think I nuked my Treesitter installation and started from scratch.

That shouldn't be the source of the issue though. Try setting `vim.o.background` to be 'light' at the end of your config and see if that does anything.

**Reply by rnl16616** [reply]
Thanks for your help and patience. You put me on the right path and you were right about treesitter - unrelated. 

The issue was surprisingly simple and obvious once it clicked - it was the order of my init.lua and specifically calling colorscheme in options.lua before onedarkpro.lua. Requiring the options.lua after onedarkpro.lua and therefore setting the colorscheme when onedarkpro was setup solved the problem.

There is one last thing I would like to do if I can prevail on your patience for a little longer... Is it possible to use white/font foreground generally as inherited from Windows Terminal but then set the NeoVim terminal to use a green font? I have not found a reliable way to set it other than to inherit it from Windows Terminal and then it makes the other foreground/fonts green too like the command line. I can't seem to separate the two things such that I can have one font white and in the terminal only it is green.

**Comment by mmirus** [comment]
@olimorris since this is coming up a lot, would you consider updating the default `StatusLine` / `StatusLineNC` highlights? https://github.com/mmirus/onedarkpro.nvim/blob/main/lua/onedarkpro/highlights/editor.lua#L91-L97

One of these might work:

- add bold style to `StatusLine` (this would mimic the default neovim styles with no theme, I believe)
- or slightly dim the fg color in `StatusLineNC`

Any change to either of these highlights would fix this problem, since neovim would no long feel like it needs to add the carets to help the user disambiguate them.

PS: this doesn't impact users if they have the `window_unfocused_color` setting enabled.

**Reply by olimorris** [reply]
Like so?

```lua
        StatusLine = { bg = config.options.transparency and "NONE" or theme.palette.bg, fg = theme.palette.fg, style = "bold" }, -- status line of current window
        StatusLineNC = {
            bg = config.options.transparency and "NONE"
                or config.options.window_unfocused_color and theme.generated.color_column
                or theme.palette.bg,
            fg = theme.palette.fg,
            style = "bold"
}
```

**Reply by mmirus** [reply]
No, just on `StatusLine`:

```lua
        StatusLine = {
            bg = config.options.transparency and "NONE" or theme.palette.bg,
            fg = theme.palette.fg,
            style = "bold",
        }, -- status line of current window
        StatusLineNC = {
            bg = config.options.transparency and "NONE"
                or config.options.window_unfocused_color and theme.generated.color_column
                or theme.palette.bg,
            fg = theme.palette.fg,
        }, -- status lines of not-current windows Note: if this is equal to "StatusLine" Vim will use "^^^" in the status line of the current window.
```

Resulting in this when not using lualine (note the bold text and lack of ^^^ in the statusline for the right-hand window, which is the active one):

![image](https://user-images.githubusercontent.com/524994/195391469-2488435d-43df-4e9a-bd72-c2c5bc01d931.png)

When you ARE using lualine, the text will not be bold, because the lualine highlights take precedence, but the ^^^ will still not appear because neovim sees that the StatusLine and StatusLineNC highlights are different. So, this would fix it for both lualine and stock statusline users, and presumably also feline, etc.

For reference, here's the same behavior visible in a `--clean` neovim:

![image](https://user-images.githubusercontent.com/524994/195392024-d6e203e4-7b13-4ba2-a3a6-62989e38cb2c.png)

(Left side is active)
